File: packages/android/crumb-core/src/main/kotlin/dev/crumb/core/CrumbReportUploader.kt
package dev.crumb.core

import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.util.concurrent.CancellationException
import java.util.concurrent.atomic.AtomicBoolean

data class CrumbUploadHttpRequest(
    val url: String,
    val method: String,
    val headers: Map<String, String>,
    val body: ByteArray?,
    val timeoutMillis: Int,
)

data class CrumbUploadHttpResponse(
    val statusCode: Int,
    val body: ByteArray = ByteArray(0),
)

fun interface CrumbUploadTransport {
    fun send(request: CrumbUploadHttpRequest): CrumbUploadHttpResponse

    fun cancel() = Unit
}

data class CrumbUploadPassResult(
    val uploadedReportCount: Int,
    val remainingReportCount: Int,
    val shouldRetry: Boolean,
    val wasCancelled: Boolean,
)

/** Performs one bounded upload pass. Scheduling and connectivity live in crumb-ui. */
class CrumbReportUploadWorker internal constructor(
    private val queue: CrumbReportQueue,
    private val settings: CrumbUploadSettings,
    private val transport: CrumbUploadTransport,
) {
    private val cancelled = AtomicBoolean(false)

    @JvmSynthetic
    fun prepareForPass() {
        cancelled.set(false)
    }

    @JvmSynthetic
    fun cancel() {
        cancelled.set(true)
        transport.cancel()
    }

    @JvmSynthetic
    fun runPass(): CrumbUploadPassResult {
        var uploadedReportCount = 0
        var shouldRetry = false
        var wasCancelled = false

        try {
            queue.recoverInterruptedUploads()
            for (report in queue.reports()) {
                try {
                    checkCancellation()
                    queue.markUploading(report.reportId)
                    val payload = queue.load(report.reportId)
                    upload(payload)
                    checkCancellation()
                    queue.remove(report.reportId)
                    uploadedReportCount += 1
                } catch (_: CancellationException) {
                    runCatching { queue.markPending(report.reportId) }
                    wasCancelled = true
                    break
                } catch (failure: CrumbUploadFailure) {
                    runCatching { queue.markFailed(report.reportId, failure.reason) }
                    if (failure.retryable) {
                        shouldRetry = true
                        break
                    }
                } catch (_: Exception) {
                    runCatching { queue.markFailed(report.reportId, "uploader.local_failure") }
                }
            }
        } catch (_: CancellationException) {
            wasCancelled = true
        } catch (_: Exception) {
            shouldRetry = true
        }

        val remainingReportCount = runCatching { queue.reports().size }.getOrDefault(1)
        return CrumbUploadPassResult(
            uploadedReportCount = uploadedReportCount,
            remainingReportCount = remainingReportCount,
            shouldRetry = shouldRetry,
            wasCancelled = wasCancelled,
        )
    }

    private fun upload(payload: CrumbQueuedReportPayload) {
        val reportId = payload.summary.reportId
        val initialized = decodeInitialization(
            sendLifecycle(
                components = listOf("sdk", "v1", "reports", "init"),
                reportId = reportId,
                operation = "init",
                body = wrapEnvelope(payload.envelope),
            ).body,
            reportId,
        )

        when (initialized.status) {
            "accepted" -> if (initialized.artifacts.isNotEmpty()) {
                throw CrumbUploadFailure("init.invalid_response", false)
            }
            "initialized" -> {
                validateTargets(initialized.artifacts, payload)
                payload.artifacts.forEach { artifact ->
                    checkCancellation()
                    val target = initialized.artifacts.first { it.id == artifact.manifest.id }
                    put(artifact.bytes, target)
                }
            }
            else -> throw CrumbUploadFailure("init.terminal_state", false)
        }

        sendLifecycle(
            components = listOf("sdk", "v1", "reports", reportId, "complete"),
            reportId = reportId,
            operation = "complete",
            body = EMPTY_OPERATION_BODY,
        )
    }

    private fun sendLifecycle(
        components: List<String>,
        reportId: String,
        operation: String,
        body: ByteArray?,
    ): CrumbUploadHttpResponse {
        val headers = linkedMapOf(
            "Authorization" to "Bearer ${settings.projectKey}",
            "Idempotency-Key" to "$reportId:$operation",
        )
        if (body != null) headers["Content-Type"] = "application/json"
        val response = send(
            CrumbUploadHttpRequest(
                url = endpoint(components),
                method = "POST",
                headers = headers,
                body = body,
                timeoutMillis = 15_000,
            ),
            operation,
        )
        if (response.statusCode !in 200..299) throw httpFailure(operation, response.statusCode)
        return response
    }

    private fun put(bytes: ByteArray, target: InitializationArtifact) {
        if (target.method != "PUT" || !validUploadUrl(target.url)) {
            throw CrumbUploadFailure("artifact.invalid_target", false)
        }
        val response = send(
            CrumbUploadHttpRequest(
                url = target.url,
                method = "PUT",
                headers = target.headers.filterKeys(::validHeaderName),
                body = bytes,
                timeoutMillis = 60_000,
            ),
            "artifact",
        )
        if (response.statusCode !in 200..299) throw httpFailure("artifact", response.statusCode)
    }

    private fun send(request: CrumbUploadHttpRequest, operation: String): CrumbUploadHttpResponse {
        checkCancellation()
        return try {
            transport.send(request).also { checkCancellation() }
        } catch (_: CancellationException) {
            throw CancellationException()
        } catch (_: Exception) {
            checkCancellation()
            throw CrumbUploadFailure("$operation.network", true)
        }
    }

    private fun decodeInitialization(bytes: ByteArray, reportId: String): InitializationResponse {
        try {
            val root = JSONObject(String(bytes, Charsets.UTF_8))
            if (root.getString("report_id") != reportId) {
                throw CrumbUploadFailure("init.invalid_response", false)
            }
            val artifacts = root.getJSONArray("artifacts")
            return InitializationResponse(
                reportId = reportId,
                status = root.getString("status"),
                artifacts = (0 until artifacts.length()).map { index ->
                    val value = artifacts.getJSONObject(index)
                    val headerObject = value.getJSONObject("headers")
                    val headers = headerObject.keys().asSequence().associateWith(headerObject::getString)
                    InitializationArtifact(
                        id = value.getString("id"),
                        uploadId = value.getString("upload_id"),
                        method = value.getString("method"),
                        url = value.getString("url"),
                        headers = headers,
                    )
                },
            )
        } catch (failure: CrumbUploadFailure) {
            throw failure
        } catch (_: Exception) {
            throw CrumbUploadFailure("init.invalid_response", false)
        }
    }

    private fun validateTargets(
        targets: List<InitializationArtifact>,
        payload: CrumbQueuedReportPayload,
    ) {
        val targetsById = targets.associateBy(InitializationArtifact::id)
        if (
            targetsById.size != targets.size ||
            targetsById.keys != payload.artifacts.map { it.manifest.id }.toSet() ||
            payload.artifacts.any { artifact ->
                targetsById[artifact.manifest.id]?.uploadId != artifact.manifest.uploadId
            }
        ) {
            throw CrumbUploadFailure("init.invalid_response", false)
        }
    }

    private fun wrapEnvelope(envelope: ByteArray): ByteArray =
        "{\"envelope\":".toByteArray() + envelope + "}".toByteArray()

    private fun endpoint(components: List<String>): String =
        settings.ingestionUrl.trimEnd('/') + "/" + components.joinToString("/")

    private fun validUploadUrl(value: String): Boolean {
        val uri = runCatching { URI(value) }.getOrNull() ?: return false
        val scheme = uri.scheme?.lowercase()
        val baseScheme = URI(settings.ingestionUrl).scheme.lowercase()
        return uri.isAbsolute && !uri.host.isNullOrBlank() && uri.userInfo == null &&
            uri.fragment == null && (scheme == "https" || (scheme == "http" && baseScheme == "http"))
    }

    private fun validHeaderName(value: String): Boolean =
        value.isNotEmpty() && value.all { it in HEADER_CHARACTERS }

    private fun httpFailure(operation: String, statusCode: Int): CrumbUploadFailure {
        val retryable = statusCode in setOf(408, 425, 429) || statusCode in 500..599 ||
            (operation == "artifact" && statusCode == 403)
        return CrumbUploadFailure("$operation.http_$statusCode", retryable)
    }

    private fun checkCancellation() {
        if (cancelled.get() || Thread.currentThread().isInterrupted) throw CancellationException()
    }

    companion object {
        private val EMPTY_OPERATION_BODY = "{}".toByteArray()

        private const val HEADER_CHARACTERS =
            "!#\$%&'*+-.^_`|~0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"

        @JvmSynthetic
        fun create(
            queue: CrumbReportQueue,
            settings: CrumbUploadSettings,
        ): CrumbReportUploadWorker = CrumbReportUploadWorker(
            queue,
            settings,
            CrumbHttpUrlConnectionUploadTransport(),
        )
    }
}

private class CrumbHttpUrlConnectionUploadTransport : CrumbUploadTransport {
    @Volatile
    private var activeConnection: HttpURLConnection? = null

    override fun send(request: CrumbUploadHttpRequest): CrumbUploadHttpResponse {
        val connection = URL(request.url).openConnection() as HttpURLConnection
        activeConnection = connection
        try {
            connection.requestMethod = request.method
            connection.connectTimeout = request.timeoutMillis
            connection.readTimeout = request.timeoutMillis
            connection.instanceFollowRedirects = false
            connection.useCaches = false
            request.headers.forEach(connection::setRequestProperty)
            request.body?.let { body ->
                connection.doOutput = true
                connection.setFixedLengthStreamingMode(body.size)
                connection.outputStream.use { it.write(body) }
            }
            val statusCode = connection.responseCode
            val input = if (statusCode >= 400) connection.errorStream else connection.inputStream
            return CrumbUploadHttpResponse(statusCode, input.readBounded())
        } finally {
            activeConnection = null
            connection.disconnect()
        }
    }

    override fun cancel() {
        activeConnection?.disconnect()
    }
}

private fun InputStream?.readBounded(maximumBytes: Int = 1_048_576): ByteArray {
    if (this == null) return ByteArray(0)
    use { stream ->
        val output = ByteArrayOutputStream()
        val buffer = ByteArray(8_192)
        while (true) {
            val count = stream.read(buffer)
            if (count < 0) break
            if (output.size() + count > maximumBytes) throw IOException("upload response exceeded limit")
            output.write(buffer, 0, count)
        }
        return output.toByteArray()
    }
}

private data class InitializationResponse(
    val reportId: String,
    val status: String,
    val artifacts: List<InitializationArtifact>,
)

private data class InitializationArtifact(
    val id: String,
    val uploadId: String,
    val method: String,
    val url: String,
    val headers: Map<String, String>,
)

private class CrumbUploadFailure(
    val reason: String,
    val retryable: Boolean,
) : IllegalStateException(reason)

File: packages/ios/Sources/CrumbCore/CrumbReportUploader.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

package struct CrumbUploadHTTPResponse: Sendable {
    package let statusCode: Int
    package let data: Data

    package init(statusCode: Int, data: Data = Data()) {
        self.statusCode = statusCode
        self.data = data
    }
}

package protocol CrumbUploadTransport: Sendable {
    func send(_ request: URLRequest) async throws -> CrumbUploadHTTPResponse
}

package struct CrumbURLSessionUploadTransport: CrumbUploadTransport {
    private let session: URLSession

    package init(session: URLSession = .shared) {
        self.session = session
    }

    package func send(_ request: URLRequest) async throws -> CrumbUploadHTTPResponse {
        let (data, response) = try await session.data(for: request)
        guard data.count <= 1_048_576,
              let response = response as? HTTPURLResponse else {
            throw CrumbUploadFailure(reason: "transport.invalid_response", retryable: true)
        }
        return CrumbUploadHTTPResponse(statusCode: response.statusCode, data: data)
    }
}

package struct CrumbUploadPassResult: Equatable, Sendable {
    package let uploadedReportCount: Int
    package let remainingReportCount: Int
    package let shouldRetry: Bool
    package let wasCancelled: Bool
}

/// Performs one bounded upload pass. Scheduling and connectivity live in CrumbUI.
package actor CrumbReportUploadWorker {
    private let queue: CrumbReportQueue
    private let settings: CrumbUploadSettings
    private let transport: any CrumbUploadTransport

    package init(
        queue: CrumbReportQueue = .shared,
        settings: CrumbUploadSettings,
        transport: any CrumbUploadTransport = CrumbURLSessionUploadTransport()
    ) {
        self.queue = queue
        self.settings = settings
        self.transport = transport
    }

    package func runPass() async -> CrumbUploadPassResult {
        var uploadedReportCount = 0
        var shouldRetry = false
        var wasCancelled = false

        do {
            try await queue.recoverInterruptedUploads()
            let reports = try await queue.reports()
            for report in reports {
                do {
                    try Task.checkCancellation()
                    try await queue.markUploading(reportID: report.reportID)
                    let payload = try await queue.load(reportID: report.reportID)
                    try await upload(payload)
                    try Task.checkCancellation()
                    try await queue.remove(reportID: report.reportID)
                    uploadedReportCount += 1
                } catch is CancellationError {
                    try? await queue.markPending(reportID: report.reportID)
                    wasCancelled = true
                    break
                } catch let failure as CrumbUploadFailure {
                    try? await queue.markFailed(reportID: report.reportID, reason: failure.reason)
                    if failure.retryable {
                        shouldRetry = true
                        break
                    }
                } catch {
                    try? await queue.markFailed(
                        reportID: report.reportID,
                        reason: "uploader.local_failure"
                    )
                }
            }
        } catch is CancellationError {
            wasCancelled = true
        } catch {
            shouldRetry = true
        }

        let remainingReports = try? await queue.reports()
        return CrumbUploadPassResult(
            uploadedReportCount: uploadedReportCount,
            remainingReportCount: remainingReports?.count ?? 1,
            shouldRetry: shouldRetry,
            wasCancelled: wasCancelled
        )
    }

    private func upload(_ payload: CrumbQueuedReportPayload) async throws {
        let reportID = payload.summary.reportID
        let initResponse = try await sendLifecycle(
            path: ["sdk", "v1", "reports", "init"],
            reportID: reportID,
            operation: "init",
            body: wrappedEnvelope(payload.envelope)
        )
        let initialized = try decodeInitialization(initResponse.data, reportID: reportID)

        switch initialized.status {
        case "accepted":
            guard initialized.artifacts.isEmpty else {
                throw CrumbUploadFailure(reason: "init.invalid_response", retryable: false)
            }
        case "initialized":
            try validateTargets(initialized.artifacts, payload: payload)
            for artifact in payload.artifacts {
                try Task.checkCancellation()
                guard let target = initialized.artifacts.first(where: {
                    $0.id == artifact.manifest.id
                }) else {
                    throw CrumbUploadFailure(reason: "init.invalid_response", retryable: false)
                }
                try await put(artifact.data, target: target)
            }
        default:
            throw CrumbUploadFailure(reason: "init.terminal_state", retryable: false)
        }

        _ = try await sendLifecycle(
            path: ["sdk", "v1", "reports", reportID, "complete"],
            reportID: reportID,
            operation: "complete",
            body: nil
        )
    }

    private func sendLifecycle(
        path: [String],
        reportID: String,
        operation: String,
        body: Data?
    ) async throws -> CrumbUploadHTTPResponse {
        var request = URLRequest(url: endpoint(path))
        request.httpMethod = "POST"
        request.timeoutInterval = 15
        request.setValue("Bearer \(settings.projectKey)", forHTTPHeaderField: "Authorization")
        request.setValue("\(reportID):\(operation)", forHTTPHeaderField: "Idempotency-Key")
        if let body {
            request.httpBody = body
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        let response = try await send(request, operation: operation)
        guard (200..<300).contains(response.statusCode) else {
            throw httpFailure(operation: operation, statusCode: response.statusCode)
        }
        return response
    }

    private func put(_ data: Data, target: InitializationResponse.Artifact) async throws {
        guard target.method == "PUT",
              let url = URL(string: target.url),
              validUploadURL(url) else {
            throw CrumbUploadFailure(reason: "artifact.invalid_target", retryable: false)
        }
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.timeoutInterval = 60
        request.httpBody = data
        for (name, value) in target.headers where validHeaderName(name) {
            request.setValue(value, forHTTPHeaderField: name)
        }
        let response = try await send(request, operation: "artifact")
        guard (200..<300).contains(response.statusCode) else {
            throw httpFailure(operation: "artifact", statusCode: response.statusCode)
        }
    }

    private func send(_ request: URLRequest, operation: String) async throws -> CrumbUploadHTTPResponse {
        do {
            return try await transport.send(request)
        } catch is CancellationError {
            throw CancellationError()
        } catch let failure as CrumbUploadFailure {
            throw failure
        } catch {
            if Task.isCancelled { throw CancellationError() }
            throw CrumbUploadFailure(reason: "\(operation).network", retryable: true)
        }
    }

    private func validateTargets(
        _ targets: [InitializationResponse.Artifact],
        payload: CrumbQueuedReportPayload
    ) throws {
        let targetIDs = targets.map(\.id)
        guard targetIDs.count == Set(targetIDs).count,
              Set(targetIDs) == Set(payload.artifacts.map(\.manifest.id)) else {
            throw CrumbUploadFailure(reason: "init.invalid_response", retryable: false)
        }
        for target in targets {
            guard let artifact = payload.artifacts.first(where: { $0.manifest.id == target.id }),
                  target.uploadID == artifact.manifest.uploadID else {
                throw CrumbUploadFailure(reason: "init.invalid_response", retryable: false)
            }
        }
    }

    private func decodeInitialization(_ data: Data, reportID: String) throws -> InitializationResponse {
        do {
            let response = try JSONDecoder().decode(InitializationResponse.self, from: data)
            guard response.reportID == reportID else {
                throw CrumbUploadFailure(reason: "init.invalid_response", retryable: false)
            }
            return response
        } catch let failure as CrumbUploadFailure {
            throw failure
        } catch {
            throw CrumbUploadFailure(reason: "init.invalid_response", retryable: false)
        }
    }

    private func wrappedEnvelope(_ envelope: Data) -> Data {
        var body = Data("{\"envelope\":".utf8)
        body.append(envelope)
        body.append(Data("}".utf8))
        return body
    }

    private func endpoint(_ components: [String]) -> URL {
        components.reduce(settings.ingestionURL) { url, component in
            url.appendingPathComponent(component)
        }
    }

    private func validUploadURL(_ url: URL) -> Bool {
        let scheme = url.scheme?.lowercased()
        let baseScheme = settings.ingestionURL.scheme?.lowercased()
        return url.host != nil
            && url.user == nil
            && url.password == nil
            && url.fragment == nil
            && (scheme == "https" || (scheme == "http" && baseScheme == "http"))
    }

    private func validHeaderName(_ value: String) -> Bool {
        let allowed = CharacterSet(
            charactersIn: "!#$%&'*+-.^_`|~0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
        )
        return !value.isEmpty && value.unicodeScalars.allSatisfy(allowed.contains)
    }

    private func httpFailure(operation: String, statusCode: Int) -> CrumbUploadFailure {
        let retryable = statusCode == 408 || statusCode == 425 || statusCode == 429
            || (500...599).contains(statusCode)
            || (operation == "artifact" && statusCode == 403)
        return CrumbUploadFailure(
            reason: "\(operation).http_\(statusCode)",
            retryable: retryable
        )
    }

    private struct InitializationResponse: Decodable, Sendable {
        let reportID: String
        let status: String
        let artifacts: [Artifact]

        enum CodingKeys: String, CodingKey {
            case reportID = "report_id"
            case status
            case artifacts
        }

        struct Artifact: Decodable, Sendable {
            let id: String
            let uploadID: String
            let method: String
            let url: String
            let headers: [String: String]

            enum CodingKeys: String, CodingKey {
                case id
                case uploadID = "upload_id"
                case method
                case url
                case headers
            }
        }
    }
}

private struct CrumbUploadFailure: Error, Sendable {
    let reason: String
    let retryable: Bool
}
